src/storage/concurrency_manager/key_handle.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use super::lock_table::LockTable;
+
+use parking_lot::Mutex;
+use std::{mem, sync::Arc};
+use tokio::sync::{Mutex as AsyncMutex, MutexGuard as AsyncMutexGuard};
+use txn_types::{Key, Lock};
+
+/// An entry in the in-memory table providing functions related to a specific
+/// key.
+pub struct KeyHandle {
+    pub key: Key,
+    table: LockTable,
+    mutex: AsyncMutex<()>,
+    lock_store: Mutex<Option<Lock>>,
+}
+
+impl KeyHandle {
+    pub fn new(key: Key, table: LockTable) -> Self {
+        KeyHandle {
+            key,
+            table,
+            mutex: AsyncMutex::new(()),
+            lock_store: Mutex::new(None),
+        }
+    }
+
+    pub async fn lock(self: Arc<Self>) -> KeyHandleGuard {
+        // Safety: `_mutex_guard` is declared before `handle_ref` in `KeyHandleGuard`.
+        // So the mutex guard will be released earlier than the `Arc<KeyHandle>`.
+        // Then we can make sure the mutex guard doesn't point to released memory.
+        let mutex_guard = unsafe { mem::transmute(self.mutex.lock().await) };
+        KeyHandleGuard {
+            _mutex_guard: mutex_guard,
+            handle: self,
+        }
+    }
+
+    pub fn with_lock<T>(&self, f: impl FnOnce(&Option<Lock>) -> T) -> T {
+        f(&*self.lock_store.lock())
+    }
+}
+
+impl Drop for KeyHandle {
+    fn drop(&mut self) {
+        self.table.remove(&self.key);
+    }
+}
+
+/// A `KeyHandle` with its mutex locked.
+pub struct KeyHandleGuard {
+    // It must be declared before `handle_ref` so it will be dropped before
+    // `handle_ref`.
+    _mutex_guard: AsyncMutexGuard<'static, ()>,
+    handle: Arc<KeyHandle>,
+}
+
+impl KeyHandleGuard {
+    pub fn key(&self) -> &Key {
+        &self.handle.key
+    }
+
+    pub fn with_lock<T>(&self, f: impl FnOnce(&mut Option<Lock>) -> T) -> T {
+        f(&mut *self.handle.lock_store.lock())
+    }
+}
+
+impl Drop for KeyHandleGuard {
+    fn drop(&mut self) {
+        // We only keep the lock in memory until the write to the underlying
+        // store finishes.
+        // The guard can be released after finishes writing.
+        // FIXME: read checks should be done before getting the snapshot
+        *self.handle.lock_store.lock() = None;
+    }
+}
```

Comment:
@youjiali1995 I follow your idea and drop the stored lock here. Then, we must do read checks before getting the snapshot. However, it is a bit difficult to do that under current coprocessor structure...

Reply:
😑